src/Data/UriPath.hs

```diff
@@ -0,0 +1,72 @@
+{-# LANGUAGE FlexibleInstances     #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE OverloadedStrings     #-}
+{-# LANGUAGE TypeFamilies          #-}
+module Data.UriPath
+    ( UriPart
+    , UriPath
+    , (</>)
+    , absoluteUriPath
+    , relativeUriPath
+    , href_
+    , onclick_
+    , HasPath(..)
+    , HasUriPart(..))
+    where
+
+import Thentos.Prelude
+import Data.String.Conversions
+
+import qualified Data.Text as ST
+import qualified Lucid
+
+newtype UriPart = SlashFreeUriPart { fromUriPart :: ST }
+
+instance IsString UriPart where
+    fromString s
+        | '/' `elem` s = error $ "UriPart.fromString: unexpected literal '/' in URI part: " <> s
```

Comment:
Better would be to statically check the string literals.

Another option would be to use URI encoding.